### DISCUSSION TITLE
UIViewRepresentable that wraps a TextKit 1 UITextView gives a blank snapshot.

### DISCUSSION THREAD

**Discussion by pixlwave** [pointfreeco/swift-snapshot-testing#872] [General]
**Describe the bug**
Hi, I'm attempting to update our project from Xcode 15.2 to Xcode 15.4 and have hit a weird bug in our snapshot tests:

We have a custom SwiftUI text view that wraps a `UITextView` that's using the old TextKit 1 layout engine for various custom layout needs. With Xcode 15.4, when the height of this view exceeds a certain threshold (~5 lines of body text), then the text is no longer visible in any snapshots. All of our snapshots are generated from Xcode Previews which don't exhibit this behaviour so it appears to only be an issue with the snapshotting.

Allowing the text view to use TextKit 2 by initialising with `usingTextLayoutManager: true` fixes the snapshots but sadly isn't a viable option for us at this time.

**To Reproduce**
A minimal reproduction case looks like so:

```swift
import SwiftUI

struct CustomText: UIViewRepresentable {
    let attributedString: AttributedString

    func makeUIView(context: Context) -> UITextView {
        // Need to use TextKit 1 for mentions
        let textView = UITextView(usingTextLayoutManager: false)
        textView.attributedText = NSAttributedString(attributedString)
        return textView
    }

    func updateUIView(_ uiView: UITextView, context: Context) {
        uiView.attributedText = NSAttributedString(attributedString)
    }

    func sizeThatFits(_ proposal: ProposedViewSize, uiView: UITextView, context: Context) -> CGSize? {
        let proposalWidth = proposal.width ?? UIView.layoutFittingExpandedSize.width
        return uiView.sizeThatFits(CGSize(width: proposalWidth, height: UIView.layoutFittingCompressedSize.height))
    }
}

// MARK: - Previews

struct CustomText_Previews: PreviewProvider {
    static var previews: some View {
        CustomText(attributedString: makeBasicAttributedString())
            .border(Color.purple)
    }
    
    static func makeBasicAttributedString() -> AttributedString {
        let string = "This is a list\n• One\n• Two\n• Three\n• Four\n• Five\n• Six\n• Seven\n• Eight\n• Nine\n• And number 10\nThat was a list"
        let attributedString = NSAttributedString(string: string)
        
        return AttributedString(attributedString)
    }
}
```

```swift
import SnapshotTesting
import XCTest
@testable import TextKitSnapshots

final class TextKitSnapshotsTests: XCTestCase {
    func testSnapshot() throws {
        let view = CustomText_Previews.previews
            .frame(width: 390, height: 844)
        assertSnapshot(of: view, as: .image)
    }
}
```

**Expected behavior**
The snapshots should continue to render as they do when generated by Xcode 15.2.

**Screenshots**
| Xcode 15.2 (iOS 17.2 Simulator) | Xcode 15.4 (iOS 17.5 Simulator) |
| - | - |
| ![Xcode 15 2](https://github.com/pointfreeco/swift-snapshot-testing/assets/6060466/4e319fed-a0ab-4892-86e6-fcde6b437462) | ![Xcode15 4](https://github.com/pointfreeco/swift-snapshot-testing/assets/6060466/5026ca54-1990-4452-b535-46a4feaea9c6) |

**Environment**
  - swift-snapshot-testing version: 1.17.1
  - Xcode: 15.4
  - Swift: 5.10
  - OS: iOS 17.5

**Additional context**
Sample Xcode project: [TextKitSnapshots.zip](https://github.com/user-attachments/files/16166612/TextKitSnapshots.zip)

**Comment by mbrandonw** [comment]
Hi @pixlwave, something in Xcode/iOS must have changed this, but I don't think it's really a bug per se with the library. For these kinds of open ended questions I think discussions are better, so I am going to move this over to the discussions tab of the repo. Perhaps someone can help you figure it out over there!

**Reply by pixlwave** [reply]
Sorry my bad @mbrandonw. I assumed this would be considered a bug in the library given the snapshots are rendered differently to the app.

**Comment by pixlwave** [comment]
Ah ha, so it seems the fix is from the sizing method. Moving the frame from the view to a device layout fixes the issue:

```swift
func testSnapshot() throws {
    let view = CustomText_Previews.previews
    assertSnapshot(of: view, as: .image(layout: .device(config: .iPhone13)))
}
```

**Reply by pixlwave** [reply]
Although, using either of `.image(layout: .sizeThatFits)` or `.image(layout: .fixed(…))` still fails even without the frame on the view. 🤔

**Reply by pixlwave** [reply]
Digging into this a bit more using a custom config on the `device` layout, it appears that it needs a non-zero `safeArea` to work. So

```swift
.device(config: .init(safeArea: .init(top: 1, left: 0, bottom: 0, right: 0),
                      size: .init(width: 390, height: 844),
                      traits: .iPhone13(.portrait)))
```

works, whereas

```swift
.device(config: .init(safeArea: .init(top: 0, left: 0, bottom: 0, right: 0),
                      size: .init(width: 390, height: 844),
                      traits: .iPhone13(.portrait)))
```
does not.

**Reply by pixlwave** [reply]
Ok so, if I comment out this line that moves the view way off screen:
https://github.com/pointfreeco/swift-snapshot-testing/blob/63d3b45dd249878a41c56274a748ca2c1c9c5230/Sources/SnapshotTesting/Common/View.swift#L986
then the original `assertSnapshot(of: view, as: .image)` works as expected.

@mbrandonw I appreciate this isn't the right fix, but I now wonder if this makes it count as a bug in the way the library handles safe area insets?
